### ISSUE TITLE
"ERROR: requires Universal Ctags to work properly" when using -public-headers option

### ISSUE THREAD

**Issue by Knitschi** [lvc/abi-dumper#12]
Hi, I am creating an abi dump file and I added the `-public-headers "/path/to/my/headers.txt"` option. The content of `headers.txt` is something like:

````
/home/knitschi/CppLibraries/CppCodeBase/Sources/libQtUtilities/QObjectUtilities.h
/home/knitschi/CppLibraries/CppCodeBase/Sources/libQtUtilities/QPrintTo.h
/home/knitschi/CppLibraries/CppCodeBase/Sources/libQtUtilities/QStringUtilities.h
````

Now the tool gives me the error

````
Detect public symbols
ERROR: requires Universal Ctags to work properly
````
which does not occur when I do not add the `-public-headers` option. However the abi-dump is created and the abi-dumper does not return with an error.

Any ideas about what went wrong?

**Comment by lvc** [comment]
Hi,

It suggests installing of the latest `ctags` from https://github.com/universal-ctags/ctags:

    sh ./autogen.sh
    ./configure --prefix=/usr
    make
    make install

It works better than the default `ctags` installed in your system.

Thank you.

**Comment by Knitschi** [comment]
That worked, thank you. The error message is gone now.

**Comment by mdavidsaver** [comment]
I'm seeing this "error" as well as of 3d6dbb6d48d10579984fb7fda2e79d3d73e10f70, but it appears to be a false positive now.

```sh
$ strace -o dumper -ff abi-dumper lib/linux-x86_64/libpvxs.so.* -o change.dump -public-headers include -lver 1 -lambda
Detect public symbols
ERROR: requires Universal Ctags to work properly
Reading debug-info
Reading v-tables
Creating ABI dump

The object ABI has been dumped to:
  change.dump
```

Despite the message,  `/usr/bin/ctags` is being exec'd.

```sh
$ grep -i ctags dumper.*
dumper.25035:stat("ctags", 0x55c7ba828478)           = -1 ENOENT (No such file or directory)
dumper.25035:stat("/bin/ctags", {st_mode=S_IFREG|0755, st_size=280176, ...}) = 0
dumper.25035:read(3, "Exuberant Ctags 5.9~svn20110310,"..., 8192) = 197
dumper.25035:write(2, "ERROR: requires Universal Ctags "..., 49) = 49
dumper.25043:execve("/usr/local/bin/ctags", ["ctags", "--version"], 0x55c7ba844990 /* 53 vars */) = -1 ENOENT (No such file or directory)
dumper.25043:execve("/usr/bin/ctags", ["ctags", "--version"], 0x55c7ba844990 /* 53 vars */) = 0
...
```

The `ctags` executable is in my `$PATH`.

```sh
$ which ctags
/usr/bin/ctags
$ ctags --version
Exuberant Ctags 5.9~svn20110310, Copyright (C) 1996-2009 Darren Hiebert
  Addresses: <dhiebert@users.sourceforge.net>, http://ctags.sourceforge.net
  Optional compiled features: +wildcards, +regex
```

Also, any chance of a post 1.1 release?  fyi. [Debian is still shipping](https://packages.debian.org/search?keywords=abi-dumper) 1.1, which doesn't seem to correctly handle my heavily C++ library.

**Comment by lvc** [comment]
> Also, any chance of a post 1.1 release? fyi. [Debian is still shipping](https://packages.debian.org/search?keywords=abi-dumper) 1.1, which doesn't seem to correctly handle my heavily C++ library.

Released 1.2.

> The ctags executable is in my $PATH.

You need universal ctags, not exuberant ctags. See https://github.com/universal-ctags/ctags

**Comment by mdavidsaver** [comment]
Thank you.

As a note, 1.2 ships with the recently released [Debian 11](https://packages.debian.org/bullseye/abi-dumper).  `ctags` is still defaulting to `ctags-exuberant` if both are installed, I think because `e` comes before `u`...  So I also do:

```
$ sudo update-alternatives --set ctags /usr/bin/ctags-universal
```

```
$ ctags --version
Universal Ctags 0.0.0, Copyright (C) 2015 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
...
$ aptitude show universal-ctags 
Package: universal-ctags                 
Version: 0+git20200824-1.1
...
```

And I can't help but notice that Debian 11 now includes a third ctags fork...

```
$ aptitude search ctags
p   arduino-ctags                                   - Arduino fork of exuberant ctags                          
v   ctags                                           -                                                          
i A exuberant-ctags                                 - build tag file indexes of source code definitions        
i   universal-ctags                                 - build tag file indexes of source code definitions        
```
